core/matrix/bccoo.cpp

```diff
@@ -0,0 +1,789 @@
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/matrix/bccoo.hpp>
+
+
+#include <algorithm>
+#include <numeric>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/precision_dispatch.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/components/absolute_array_kernels.hpp"
+#include "core/components/fill_array_kernels.hpp"
+#include "core/matrix/bccoo_aux_structs.hpp"
+#include "core/matrix/bccoo_helper.hpp"
+#include "core/matrix/bccoo_kernels.hpp"
+
+
+namespace gko {
+namespace matrix {
+namespace bccoo {
+
+
+GKO_REGISTER_OPERATION(get_default_block_size, bccoo::get_default_block_size);
+GKO_REGISTER_OPERATION(get_default_compression, bccoo::get_default_compression);
+GKO_REGISTER_OPERATION(spmv, bccoo::spmv);
+GKO_REGISTER_OPERATION(advanced_spmv, bccoo::advanced_spmv);
+GKO_REGISTER_OPERATION(spmv2, bccoo::spmv2);
+GKO_REGISTER_OPERATION(advanced_spmv2, bccoo::advanced_spmv2);
+GKO_REGISTER_OPERATION(convert_to_bccoo, bccoo::convert_to_bccoo);
+GKO_REGISTER_OPERATION(convert_to_next_precision,
+                       bccoo::convert_to_next_precision);
+GKO_REGISTER_OPERATION(convert_to_coo, bccoo::convert_to_coo);
+GKO_REGISTER_OPERATION(convert_to_csr, bccoo::convert_to_csr);
+GKO_REGISTER_OPERATION(convert_to_dense, bccoo::convert_to_dense);
+GKO_REGISTER_OPERATION(extract_diagonal, bccoo::extract_diagonal);
+GKO_REGISTER_OPERATION(compute_absolute_inplace,
+                       bccoo::compute_absolute_inplace);
+GKO_REGISTER_OPERATION(compute_absolute, bccoo::compute_absolute);
+GKO_REGISTER_OPERATION(mem_size_bccoo, bccoo::mem_size_bccoo);
+GKO_REGISTER_OPERATION(fill_array, components::fill_array);
+GKO_REGISTER_OPERATION(inplace_absolute_array,
+                       components::inplace_absolute_array);
+GKO_REGISTER_OPERATION(outplace_absolute_array,
+                       components::outplace_absolute_array);
+
+
+}  // namespace bccoo
+
+
+template <typename ValueType, typename IndexType>
+void Bccoo<ValueType, IndexType>::apply_impl(const LinOp* b, LinOp* x) const
+{
+    // This routine doesn't make sense for non initialized objects
+    GKO_ASSERT(this->is_initialized);
```

Comment:
But I guess you didn't push the changes? This is still in the old state.

Reply:
Not yet. I wanted to wait until the end of the review, to avoid the reception of automatic messages from github.
But I am going to do it now to make easier your review.